utils/weighted_random.go

```diff
@@ -0,0 +1,155 @@
+package utils
+
+import (
+	"fmt"
+	"hash/fnv"
+	"math/rand"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+)
+
+var reflectInt = reflect.TypeOf(int(1))
+var reflectStr = reflect.TypeOf("a")
+var reflectFloat = reflect.TypeOf(float64(1))
+
+type WeightedRandom interface {
+	Get() interface{}
+	GetWithSeed(seed string) (interface{}, error)
+}
+
+type Entry struct {
+	Item   interface{}
+	Weight float32
+}
+
+type internalEntry struct {
+	entry        Entry
+	currentTotal float32
+}
+
+// WeightedRandom selects elements randomly from the list taking into account individual weights.
+// Weight has to be assigned between 0 and 1.
+// Support deterministic results given a particular seed and sortKey
+type weightedRandomImpl struct {
+	entries     []internalEntry
+	totalWeight float32
+	mux         sync.Mutex
+}
+
+func validateEntries(entries []Entry, sortKey string) error {
+	if len(entries) == 0 {
+		return fmt.Errorf("entries is empty")
+	}
+	for _, entry := range entries {
+		if entry.Weight < 0 || entry.Weight > float32(1) {
+			return fmt.Errorf("invalid weight %f", entry.Weight)
+		}
+
+		item := reflect.ValueOf(entry.Item)
+		f := reflect.Indirect(item).FieldByName(sortKey)
+		if !f.IsValid() {
+			return fmt.Errorf("invalid sort key")
+		}
+		switch f.Type() {
+		case reflectInt:
+		case reflectStr:
+		case reflectFloat:
+			continue
+		default:
+			return fmt.Errorf("unsupported type")
+		}
+	}
+	return nil
+}
+
+// Given a list of entries and sortKey, return WeightedRandom
+// The sortKey indicates the field in the object to be used for sorting.
+// This enables deterministic results for same seed and sortKey
+func NewWeightedRandom(entries []Entry, sortKey string) (WeightedRandom, error) {
```

Comment:
Now that I've actually read through the code, I don't think there is any reason for this class to care about sorting to begin with... It's totally fair to assume the slice you receive is sorted in whatever order the caller cares about.

If you want to implement a generic sorter util, that's fine... though I wouldn't do it this way.. it's not the Go-way :)... I would either require a list of objects that implement `type Comparable interface { CompareTo(other Comparable) }` or leave it as an interface{} and code-generate multiple sorting algorithms for different types...

Reply:
Good idea. 

If we want the results to be deterministic - given a list of same weights and seed, we want the same result. Will go with the Comparable option. Looks much better than current state.